File: lib/rgfa/line/edge/gfa1/other.rb
RGFA::Line::Edge::GFA1 ||= Module.new

module RGFA::Line::Edge::GFA1::Other

  # @param oriented_segment [RGFA::OrientedLine]
  #   one of the two oriented segments of the line
  # @return [RGFA::OrientedLine] the other oriented segment
  # @raise [RGFA::NotFoundError] if segment_end is not a segment end of the line
  def other_oriented_segment(oriented_segment)
    if (oriented_from == oriented_segment)
      return oriented_to
    elsif (oriented_to == oriented_segment)
      return oriented_from
    else
      raise RGFA::NotFoundError,
        "Oriented segment '#{oriented_segment.inspect}' not found\n"+
        "Line: #{self}"
    end
  end

  # The other segment of a connection line
  # @param segment [RGFA::Line::Segment::GFA1, Symbol] segment name or instance
  # @raise [RGFA::NotFoundError]
  #   if segment is not involved in the connection
  # @return [Symbol] the name or instance of the other segment of the connection
  #   if circular, then +segment+
  def other(segment)
    segment_name = segment.to_sym
    if segment_name == from.to_sym
      to
    elsif segment_name == to.to_sym
      from
    else
      raise RGFA::NotFoundError,
        "Line #{self} does not involve segment #{segment_name}"
    end
  end

end

File: gfapy/line/edge/gfa1/other.py
import gfapy

class Other:

  def other_oriented_segment(self, oriented_segment, tolerant = False):
    """
    Parameters
    ----------
    oriented_segment : gfapy.OrientedLine
      One of the two oriented segments of the line.

    Returns
    -------
    gfapy.OrientedLine
      The other oriented segment.

    Raises
    ------
    gfapy.NotFoundError
      If segment_end is not a segment end of the line.
    """
    if (self.oriented_from == oriented_segment):
      return self.oriented_to
    elif (self.oriented_to == oriented_segment):
      return self.oriented_from
    elif tolerant:
      return None
    else:
      raise gfapy.NotFoundError(
        "Oriented segment '{}' not found\n".format(repr(oriented_segment))+
        "Line: {}".format(self))

  def other(self, segment, tolerant = False):
    """
    The other segment of a connection line.

    Parameters
    ----------
    segment : gfapy.line.segment.GFA1 or str
      Segment name or instance.

    Raises
    ------
    gfapy.NotFoundError
      If segment is not involved in the connection.

    Returns
    -------
    str
      The name or instance of the other segment of the connection.
      If circular, then **segment**.
    """
    segment_name = str(segment)
    if segment_name == str(self.from_segment):
      return self.to_segment
    elif segment_name == str(self.to_segment):
      return self.from_segment
    elif tolerant:
      return None
    else:
      raise gfapy.NotFoundError(
        "Line {} does not involve segment {}".format(self, segment_name))
